prov/efa/src/efa_av.c

```diff
@@ -330,6 +337,7 @@ int efa_av_insert_addr(struct efa_av *av, struct efa_ep_addr *addr,
 			"\n", smr_name, *(uint64_t *)addr, *fi_addr, shm_fiaddr);
 
 		assert(shm_fiaddr < EFA_SHM_MAX_AV_COUNT);
+		av->shm_used++;
```

Comment:
I thought there is a way to check av size easily, apparently these is not. Instead of introducing a new variable ```shm_used```, I think we just check ```shm_fiaddr``` here and if it is larger than ``` EFA_SHM_MAX_AV_COUNT```, do the error handling thing.

BTW, why do we introduce  ```rxr_env.shm_av_size``` in the first place?

Reply:
I thought about using shm_fiaddr. But that would assume we get the shm_fiaddr in order. Also, we have another variable `used` to track of # fi_address in efa av. 
So I applied the same method here and feel it's safer.

We introduced `rxr_env.shm_av_size` in case the users run in a system where there're more than 128 cores, so they can change it via runtime env. instead of recompiling the code.

Reply:
Currently in master we have
```
#define EFA_SHM_MAX_AV_COUNT       (256)
```
I think at some point we changed 128 to 256. Because it is 256 now, I do not see the point of have ```rxr_env.shm_av_size```.

Note ```EFA_SHM_MAX_AV_COUNT``` was used to define the ```efa_av->shm_rdm_addr_map```, so we cannot handle the case ```shm_fiaddr``` that is larger or equal to ```EFA_SHM_MAV_AV_COUNT```, so I think we need an error handling here.

Reply:
Hmm. I don't think we changed EFA_SHM_MAX_AV_COUNT before. EFA_SHM_MAX_AV_COUNT is used to track shm's SMR_MAX_PEERS, as it's not exposed to the upper layer. 
And rxr_env.shm_av_size is used for the user to adjust, but it has to be within the range of EFA_SHM_MAX_AV_COUNT. We have the following logic when openning shm AV in master to ensure the # av_entry won't be larger than rxr_env.shm_av_size and EFA_SHM_MAX_AV_COUNT as well.
```
/*
 * shm av supports maximum 256 entries
 * Reset the count to 128 to reduce memory footprint and satisfy
 * the need of the instances with more CPUs.
 */
if (rxr_env.shm_av_size > EFA_SHM_MAX_AV_COUNT) {
        ret = -FI_ENOMEM;
        goto err_close_rdm_av;
}
av_attr.count = rxr_env.shm_av_size;
assert(av_attr.type == FI_AV_TABLE);
ret = fi_av_open(efa_domain->shm_domain, &av_attr,
                &av->shm_rdm_av, context);
if (ret)
        goto err_close_rdm_av;

for (i = 0; i < EFA_SHM_MAX_AV_COUNT; ++i)
        av->shm_rdm_addr_map[i] = FI_ADDR_UNSPEC;
```

And because the variable shm_used is introduced in this PR to check if it's equal or larger than rxr_env.shm_av_size, I don't think the case you mentioned will happen.

Please let me know if I was missing something.

Reply:
Thank you! I understand the ```rxr_env.shm_av_size``` part now. 
I think we will have to handle ```shm_fiaddr >= EFA_SHM_MAX_AV_COUNT``` anyway, so it is unnecessary to introduce ```shm_used```

Reply:
And one more benefit of using `shm_used` to check is that we don't need to do the unnecessary fi_av_insert if we already know the limit has been reached.
```
if (av->shm_used >= rxr_env.shm_av_size) {
        ... error handling
}
```

Reply:
Does ```av->shm_used < rxr_env.shm_av_size``` guarantee that ```shm_fiaddr``` returned from ```fi_av_insert``` must be less than ```EFA_MAX_AV_COUNT```?